File: docs/mpt-spec/python/merkle_patricia_verifier.py
#!/usr/bin/env python3
"""
Merkle Patricia Trie Inclusion Proof Verifier

This module provides verification of Merkle Patricia inclusion proofs
compatible with the Scala implementation in the Metakit project.
Uses RFC 8785 compliant JSON canonicalization.
"""

import hashlib
import json
from typing import Dict, List, Tuple, Optional, Union, Any
from dataclasses import dataclass
from enum import Enum
import rfc8785


class VerificationError(Exception):
    """Base class for verification errors"""
    pass


class InvalidWitness(VerificationError):
    """Raised when witness structure is invalid"""
    pass


class InvalidPath(VerificationError):
    """Raised when path is invalid or not found"""
    pass


class InvalidNodeCommitment(VerificationError):
    """Raised when node commitment verification fails"""
    pass


class NodeType(Enum):
    """Node type identifiers"""
    LEAF = "Leaf"
    BRANCH = "Branch"
    EXTENSION = "Extension"


class NodePrefix:
    """Binary prefixes for node types during hashing"""
    LEAF = bytes([0])
    BRANCH = bytes([1])
    EXTENSION = bytes([2])


class Nibble:
    """Utilities for handling nibbles (4-bit values)"""
    
    @staticmethod
    def from_hex_string(hex_string: str) -> List[int]:
        """Convert hex string to list of nibbles"""
        nibbles = []
        for char in hex_string:
            try:
                nibbles.append(int(char, 16))
            except ValueError:
                raise ValueError(f"Invalid hex character: {char}")
        return nibbles
    
    @staticmethod
    def from_hash(hash_str: str) -> List[int]:
        """Convert hash string to nibbles"""
        return Nibble.from_hex_string(hash_str)
    
    @staticmethod
    def sequence_equals(a: List[int], b: List[int]) -> bool:
        """Check if two nibble sequences are equal"""
        return a == b
    
    @staticmethod
    def sequence_drop(sequence: List[int], count: int) -> List[int]:
        """Drop first count elements from sequence"""
        return sequence[count:]


class JsonBinaryHasher:
    """Hash computation for JSON objects with binary prefixes using RFC 8785"""
    
    @staticmethod
    def compute_digest(json_object: Dict, prefix_bytes: bytes = b"") -> str:
        """Compute SHA-256 hash of JSON object with optional prefix"""
        # Use RFC 8785 canonical JSON serialization (returns bytes)
        canonical_json_bytes = rfc8785.dumps(json_object)
        
        # Combine prefix and JSON bytes
        combined = prefix_bytes + canonical_json_bytes
        
        # Compute SHA-256 hash
        hash_obj = hashlib.sha256()
        hash_obj.update(combined)
        return hash_obj.hexdigest()


@dataclass
class VerificationResult:
    """Result of verification attempt"""
    success: bool
    error: Optional[VerificationError] = None
    remaining_path: Optional[List[int]] = None
    child_digest: Optional[str] = None


class MerklePatriciaVerifier:
    """Verifier for Merkle Patricia inclusion proofs"""
    
    def __init__(self, root_hash: str):
        """
        Initialize verifier with root hash
        
        Args:
            root_hash: The root hash of the Merkle Patricia Trie
        """
        self.root_hash = root_hash
    
    def verify(self, proof: Dict) -> VerificationResult:
        """
        Verify a Merkle Patricia inclusion proof
        
        Args:
            proof: Dictionary containing 'path' and 'witness' fields
            
        Returns:
            VerificationResult indicating success or failure
        """
        try:
            # Parse the path as nibbles
            path_nibbles = Nibble.from_hash(proof['path'])
            
            # Process witness in reverse order (from root to leaf)
            witness = list(reversed(proof['witness']))
            
            # Start verification from root
            current_digest = self.root_hash
            remaining_path = path_nibbles
            index = 0
            
            while index < len(witness):
                commitment = witness[index]
                
                if 'type' not in commitment or 'contents' not in commitment:
                    raise InvalidWitness("Invalid commitment structure")
                
                node_type = commitment['type']
                contents = commitment['contents']
                
                if node_type == NodeType.LEAF.value:
                    # Leaf must be the final node
                    if index != len(witness) - 1:
                        raise InvalidWitness("Leaf must be the final commitment")
                    return self._verify_leaf(contents, current_digest, remaining_path)
                
                elif node_type == NodeType.EXTENSION.value:
                    result = self._verify_extension(contents, current_digest, remaining_path)
                    if not result.success:
                        return result
                    current_digest = contents['childDigest']
                    remaining_path = result.remaining_path
                    index += 1
                
                elif node_type == NodeType.BRANCH.value:
                    result = self._verify_branch(contents, current_digest, remaining_path)
                    if not result.success:
                        return result
                    current_digest = result.child_digest
                    remaining_path = result.remaining_path
                    index += 1
                
                else:
                    raise InvalidWitness(f"Unknown commitment type: {node_type}")
            
            raise InvalidWitness("Proof verification incomplete")
            
        except VerificationError as e:
            return VerificationResult(success=False, error=e)
        except Exception as e:
            return VerificationResult(
                success=False,
                error=InvalidWitness(f"Verification failed: {str(e)}")
            )
    
    def _verify_leaf(self, node_commit: Dict, current_digest: str, 
                     remaining_path: List[int]) -> VerificationResult:
        """Verify a leaf commitment"""
        # Compute the expected digest for this leaf commitment
        commitment_json = {
            'remaining': node_commit['remaining'],
            'dataDigest': node_commit['dataDigest']
        }
        
        computed_digest = JsonBinaryHasher.compute_digest(
            commitment_json, NodePrefix.LEAF
        )
        
        # Verify digest matches
        if computed_digest != current_digest:
            return VerificationResult(
                success=False,
                error=InvalidNodeCommitment("Invalid leaf commitment digest")
            )
        
        # Parse remaining path from commitment
        commitment_remaining = Nibble.from_hex_string(node_commit['remaining'])
        
        # Verify remaining path matches
        if not Nibble.sequence_equals(remaining_path, commitment_remaining):
            return VerificationResult(
                success=False,
                error=InvalidNodeCommitment("Path mismatch at leaf")
            )
        
        return VerificationResult(success=True)
    
    def _verify_extension(self, node_commit: Dict, current_digest: str,
                          remaining_path: List[int]) -> VerificationResult:
        """Verify an extension commitment"""
        # Compute the expected digest for this extension commitment
        commitment_json = {
            'shared': node_commit['shared'],
            'childDigest': node_commit['childDigest']
        }
        
        computed_digest = JsonBinaryHasher.compute_digest(
            commitment_json, NodePrefix.EXTENSION
        )
        
        # Verify digest matches
        if computed_digest != current_digest:
            return VerificationResult(
                success=False,
                error=InvalidNodeCommitment("Invalid extension commitment digest")
            )
        
        # Parse shared path
        shared_nibbles = Nibble.from_hex_string(node_commit['shared'])
        
        # Verify path starts with shared prefix
        for i, nibble in enumerate(shared_nibbles):
            if i >= len(remaining_path) or remaining_path[i] != nibble:
                return VerificationResult(
                    success=False,
                    error=InvalidPath("Path does not match extension shared prefix")
                )
        
        # Update remaining path
        new_remaining_path = Nibble.sequence_drop(remaining_path, len(shared_nibbles))
        
        return VerificationResult(
            success=True,
            remaining_path=new_remaining_path
        )
    
    def _verify_branch(self, node_commit: Dict, current_digest: str,
                       remaining_path: List[int]) -> VerificationResult:
        """Verify a branch commitment"""
        if not remaining_path:
            return VerificationResult(
                success=False,
                error=InvalidPath("No remaining path at branch")
            )
        
        # Get next nibble in path
        next_nibble = remaining_path[0]
        nibble_key = format(next_nibble, 'x')
        
        # Check if branch has path for this nibble
        if 'pathsDigest' not in node_commit or nibble_key not in node_commit['pathsDigest']:
            return VerificationResult(
                success=False,
                error=InvalidPath(f"Path not found in branch: {nibble_key}")
            )
        
        # Compute the expected digest for this branch commitment
        commitment_json = {
            'pathsDigest': node_commit['pathsDigest']
        }
        
        computed_digest = JsonBinaryHasher.compute_digest(
            commitment_json, NodePrefix.BRANCH
        )
        
        # Verify digest matches
        if computed_digest != current_digest:
            return VerificationResult(
                success=False,
                error=InvalidNodeCommitment("Invalid branch commitment digest")
            )
        
        # Get child digest for next step
        child_digest = node_commit['pathsDigest'][nibble_key]
        
        # Update remaining path
        new_remaining_path = remaining_path[1:]
        
        return VerificationResult(
            success=True,
            child_digest=child_digest,
            remaining_path=new_remaining_path
        )


def verify_proof_from_file(filepath: str) -> VerificationResult:
    """
    Load and verify a proof from a JSON file
    
    Args:
        filepath: Path to JSON file containing proof and root hash
        
    Returns:
        VerificationResult
    """
    with open(filepath, 'r') as f:
        data = json.load(f)
    
    verifier = MerklePatriciaVerifier(data['rootHash'])
    return verifier.verify(data['proof'])


if __name__ == "__main__":
    # Example usage
    import sys
    import os
    
    # Try to load test proof if available
    test_file = os.path.join(os.path.dirname(__file__), '..', 'test-proof.json')
    if os.path.exists(test_file):
        print("Verifying test proof...")
        result = verify_proof_from_file(test_file)
        if result.success:
            print("✓ Proof is valid!")
        else:
            print(f"✗ Proof is invalid: {result.error}")
            sys.exit(1)
    else:
        print("Test proof file not found")

File: docs/mpt-spec/js/merkle-patricia-verifier.js
const crypto = require('crypto');
const canonicalize = require('canonicalize');

/**
 * Nibble utilities - handles hex character conversions
 */
class Nibble {
  static fromHexString(hexString) {
    const nibbles = [];
    for (const char of hexString) {
      const value = parseInt(char, 16);
      if (isNaN(value)) {
        throw new Error(`Invalid hex character: ${char}`);
      }
      nibbles.push(value);
    }
    return nibbles;
  }

  static fromHash(hash) {
    return Nibble.fromHexString(hash);
  }

  static sequenceEquals(a, b) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) return false;
    }
    return true;
  }

  static sequenceDrop(sequence, count) {
    return sequence.slice(count);
  }
}

/**
 * Hash computation for JSON objects with binary prefixes
 */
class JsonBinaryHasher {
  static computeDigest(jsonObject, prefixBytes = []) {
    // Use RFC 8785 compliant canonicalization
    const canonicalJson = canonicalize(jsonObject);
    const jsonBytes = Buffer.from(canonicalJson, 'utf8');
    
    // Combine prefix and JSON bytes
    const combined = Buffer.concat([
      Buffer.from(prefixBytes),
      jsonBytes
    ]);
    
    // Compute SHA-256 hash
    const hash = crypto.createHash('sha256');
    hash.update(combined);
    return hash.digest('hex');
  }
}

/**
 * Node type prefixes for hashing
 */
const NodePrefix = {
  LEAF: [0],
  BRANCH: [1],
  EXTENSION: [2]
};

/**
 * Verification error types
 */
class MerklePatriciaVerificationError extends Error {
  constructor(message, type) {
    super(message);
    this.name = 'MerklePatriciaVerificationError';
    this.type = type;
  }
}

class InvalidWitness extends MerklePatriciaVerificationError {
  constructor(message) {
    super(message, 'InvalidWitness');
  }
}

class InvalidPath extends MerklePatriciaVerificationError {
  constructor(message) {
    super(message, 'InvalidPath');
  }
}

class InvalidNodeCommitment extends MerklePatriciaVerificationError {
  constructor(message) {
    super(message, 'InvalidNodeCommitment');
  }
}

/**
 * Merkle Patricia Verifier - verifies sealed inclusion-or-absence proofs
 *
 * This is THE reference implementation for the sealed
 * `{type: "Inclusion"|"Absence", path, witness}` wire format produced by the
 * Scala `MerklePatriciaProver.provePath` / served by `GET /committed/proof/<key>`.
 * Legacy un-tagged `{path, witness}` proofs are accepted and treated as
 * inclusion (the tagged Inclusion encoding is byte-identical plus the tag).
 */
class MerklePatriciaVerifier {
  constructor(rootHash) {
    this.rootHash = rootHash;
  }

  /**
   * Verify a sealed Merkle Patricia proof, dispatching on the proof-level
   * 'type' tag:
   *   - "Inclusion" (or no tag, the legacy shape) -> inclusion verification
   *   - "Absence" -> absence verification (same fold, different terminal assertion)
   * @param {Object} proof - { type?, path, witness }
   * @returns {Object} - { success: boolean, error?: Error }
   */
  verify(proof) {
    const tag = proof ? proof.type : undefined;
    if (tag === 'Absence') {
      return this.verifyAbsence(proof);
    }
    if (tag === undefined || tag === 'Inclusion') {
      return this.verifyInclusion(proof);
    }
    return { success: false, error: new InvalidWitness(`Unknown proof type: ${tag}`) };
  }

  /**
   * Verify a Merkle Patricia inclusion proof
   * @param {Object} proof - The inclusion proof with 'path' and 'witness' fields
   * @returns {Object} - { success: boolean, error?: Error }
   */
  verifyInclusion(proof) {
    try {
      // Parse the path as nibbles
      const pathNibbles = Nibble.fromHash(proof.path);
      
      // Process witness in reverse order (from root to leaf)
      const witness = [...proof.witness].reverse();
      
      // Start verification from root
      let currentDigest = this.rootHash;
      let remainingPath = pathNibbles;
      let index = 0;

      while (index < witness.length) {
        const commitment = witness[index];
        
        if (!commitment.type || !commitment.contents) {
          throw new InvalidWitness('Invalid commitment structure');
        }

        switch (commitment.type) {
          case 'Leaf':
            return this.verifyLeaf(commitment.contents, currentDigest, remainingPath, index === witness.length - 1);
            
          case 'Extension':
            const extResult = this.verifyExtension(commitment.contents, currentDigest, remainingPath);
            if (!extResult.success) return extResult;
            currentDigest = commitment.contents.childDigest;
            remainingPath = extResult.remainingPath;
            index++;
            break;
            
          case 'Branch':
            const branchResult = this.verifyBranch(commitment.contents, currentDigest, remainingPath);
            if (!branchResult.success) return branchResult;
            currentDigest = branchResult.childDigest;
            remainingPath = branchResult.remainingPath;
            index++;
            break;
            
          default:
            throw new InvalidWitness(`Unknown commitment type: ${commitment.type}`);
        }
      }

      throw new InvalidWitness('Proof verification incomplete');

    } catch (error) {
      if (error instanceof MerklePatriciaVerificationError) {
        return { success: false, error };
      }
      return { success: false, error: new InvalidWitness(`Verification failed: ${error.message}`) };
    }
  }

  /**
   * Verify a Merkle Patricia ABSENCE proof: NO value is committed at proof.path
   * under the root. Mirrors the Scala verifier's confirmAbsence: the witness
   * replays the SAME root-first fold as inclusion (digest recomputation per
   * commitment with the node-type prefix, child-digest threading, nibble
   * consumption); the terminal (deepest) commitment must hash to the digest the
   * fold reached AND structurally refuse the next step:
   *
   *   - Branch:    the remaining path is EMPTY (this MPT's branches carry no
   *                value slot, so a path ending at a branch is necessarily
   *                absent) OR pathsDigest lacks the next nibble;
   *   - Extension: 'shared' is NOT a prefix of the remaining path (divergence
   *                mid-edge, including a remaining path shorter than the edge);
   *   - Leaf:      'remaining' differs from the remaining path (a different key
   *                occupies the position).
   *
   * A terminal that could continue (or a Leaf that MATCHES) proves nothing.
   * @param {Object} proof - The absence proof with 'path' and 'witness' fields
   * @returns {Object} - { success: boolean, error?: Error }
   */
  verifyAbsence(proof) {
    try {
      const pathNibbles = Nibble.fromHash(proof.path);
      const witness = [...proof.witness].reverse();

      if (witness.length === 0) {
        return { success: false, error: new InvalidWitness('Empty witness') };
      }

      let currentDigest = this.rootHash;
      let remainingPath = pathNibbles;

      for (let index = 0; index < witness.length; index++) {
        const commitment = witness[index];

        if (!commitment.type || !commitment.contents) {
          throw new InvalidWitness('Invalid commitment structure');
        }

        // The deepest commitment is the divergence terminal.
        if (index === witness.length - 1) {
          return this.verifyAbsenceTerminal(commitment, currentDigest, remainingPath);
        }

        switch (commitment.type) {
          case 'Extension': {
            const extResult = this.verifyExtension(commitment.contents, currentDigest, remainingPath);
            if (!extResult.success) return extResult;
            currentDigest = commitment.contents.childDigest;
            remainingPath = extResult.remainingPath;
            break;
          }

          case 'Branch': {
            const branchResult = this.verifyBranch(commitment.contents, currentDigest, remainingPath);
            if (!branchResult.success) return branchResult;
            currentDigest = branchResult.childDigest;
            remainingPath = branchResult.remainingPath;
            break;
          }

          case 'Leaf':
            // A leaf has no child to continue into - it can only be terminal.
            return { success: false, error: new InvalidWitness('Leaf before end of witness') };

          default:
            throw new InvalidWitness(`Unknown commitment type: ${commitment.type}`);
        }
      }

      throw new InvalidWitness('Proof verification incomplete');

    } catch (error) {
      if (error instanceof MerklePatriciaVerificationError) {
        return { success: false, error };
      }
      return { success: false, error: new InvalidWitness(`Verification failed: ${error.message}`) };
    }
  }

  /**
   * The absence terminal assertion: bind the terminal commitment to the digest
   * the fold reached (same prefixed hashing as every other step), then require
   * it to structurally refuse the next step of the remaining path.
   */
  verifyAbsenceTerminal(commitment, currentDigest, remainingPath) {
    let commitmentJson;
    let prefix;

    switch (commitment.type) {
      case 'Leaf':
        commitmentJson = {
          remaining: commitment.contents.remaining,
          dataDigest: commitment.contents.dataDigest
        };
        prefix = NodePrefix.LEAF;
        break;

      case 'Branch':
        commitmentJson = {
          pathsDigest: commitment.contents.pathsDigest
        };
        prefix = NodePrefix.BRANCH;
        break;

      case 'Extension':
        commitmentJson = {
          shared: commitment.contents.shared,
          childDigest: commitment.contents.childDigest
        };
        prefix = NodePrefix.EXTENSION;
        break;

      default:
        return { success: false, error: new InvalidWitness(`Unknown commitment type: ${commitment.type}`) };
    }

    const computedDigest = JsonBinaryHasher.computeDigest(commitmentJson, prefix);
    if (computedDigest !== currentDigest) {
      return { success: false, error: new InvalidNodeCommitment('Invalid terminal commitment digest') };
    }

    let witnessesAbsence;
    if (commitment.type === 'Branch') {
      const pathsDigest = commitment.contents.pathsDigest || {};
      witnessesAbsence = remainingPath.length === 0 || !pathsDigest[remainingPath[0].toString(16)];
    } else if (commitment.type === 'Extension') {
      const sharedNibbles = Nibble.fromHexString(commitment.contents.shared);
      let sharedIsPrefix = sharedNibbles.length <= remainingPath.length;
      for (let i = 0; sharedIsPrefix && i < sharedNibbles.length; i++) {
        if (remainingPath[i] !== sharedNibbles[i]) sharedIsPrefix = false;
      }
      witnessesAbsence = !sharedIsPrefix;
    } else {
      const commitmentRemaining = Nibble.fromHexString(commitment.contents.remaining);
      witnessesAbsence = !Nibble.sequenceEquals(remainingPath, commitmentRemaining);
    }

    if (!witnessesAbsence) {
      return { success: false, error: new InvalidPath('Terminal commitment does not witness absence of the path') };
    }

    return { success: true };
  }

  verifyLeaf(nodeCommit, currentDigest, remainingPath, isFinal) {
    // Verify this is the final node in the witness
    if (!isFinal) {
      return { success: false, error: new InvalidWitness('Leaf must be the final commitment') };
    }

    // Compute the expected digest for this leaf commitment
    const commitmentJson = {
      remaining: nodeCommit.remaining,
      dataDigest: nodeCommit.dataDigest
    };
    
    const computedDigest = JsonBinaryHasher.computeDigest(commitmentJson, NodePrefix.LEAF);
    
    // Verify digest matches
    if (computedDigest !== currentDigest) {
      return { success: false, error: new InvalidNodeCommitment('Invalid leaf commitment digest') };
    }
    
    // Parse remaining path from commitment
    const commitmentRemaining = Nibble.fromHexString(nodeCommit.remaining);
    
    // Verify remaining path matches
    if (!Nibble.sequenceEquals(remainingPath, commitmentRemaining)) {
      return { success: false, error: new InvalidNodeCommitment('Path mismatch at leaf') };
    }
    
    return { success: true };
  }

  verifyExtension(nodeCommit, currentDigest, remainingPath) {
    // Compute the expected digest for this extension commitment
    const commitmentJson = {
      shared: nodeCommit.shared,
      childDigest: nodeCommit.childDigest
    };
    
    const computedDigest = JsonBinaryHasher.computeDigest(commitmentJson, NodePrefix.EXTENSION);
    
    // Verify digest matches
    if (computedDigest !== currentDigest) {
      return { success: false, error: new InvalidNodeCommitment('Invalid extension commitment digest') };
    }
    
    // Parse shared path
    const sharedNibbles = Nibble.fromHexString(nodeCommit.shared);
    
    // Verify path starts with shared prefix
    for (let i = 0; i < sharedNibbles.length; i++) {
      if (i >= remainingPath.length || remainingPath[i] !== sharedNibbles[i]) {
        return { success: false, error: new InvalidPath('Path does not match extension shared prefix') };
      }
    }
    
    // Update remaining path
    const newRemainingPath = Nibble.sequenceDrop(remainingPath, sharedNibbles.length);
    
    return { 
      success: true,
      remainingPath: newRemainingPath
    };
  }

  verifyBranch(nodeCommit, currentDigest, remainingPath) {
    if (remainingPath.length === 0) {
      return { success: false, error: new InvalidPath('No remaining path at branch') };
    }
    
    // Get next nibble in path
    const nextNibble = remainingPath[0];
    const nibbleKey = nextNibble.toString(16);
    
    // Check if branch has path for this nibble
    if (!nodeCommit.pathsDigest || !nodeCommit.pathsDigest[nibbleKey]) {
      return { success: false, error: new InvalidPath(`Path not found in branch: ${nibbleKey}`) };
    }
    
    // Compute the expected digest for this branch commitment
    const commitmentJson = {
      pathsDigest: nodeCommit.pathsDigest
    };
    
    const computedDigest = JsonBinaryHasher.computeDigest(commitmentJson, NodePrefix.BRANCH);
    
    // Verify digest matches
    if (computedDigest !== currentDigest) {
      return { success: false, error: new InvalidNodeCommitment('Invalid branch commitment digest') };
    }
    
    // Get child digest for next step
    const childDigest = nodeCommit.pathsDigest[nibbleKey];
    
    // Update remaining path
    const newRemainingPath = remainingPath.slice(1);
    
    return {
      success: true,
      childDigest,
      remainingPath: newRemainingPath
    };
  }
}

module.exports = {
  MerklePatriciaVerifier,
  MerklePatriciaVerificationError,
  InvalidWitness,
  InvalidPath,
  InvalidNodeCommitment,
  JsonBinaryHasher,
  Nibble
};